Resume Filtering

apappu9 opened the pull request:
Closes #3 

* Adds a GetFilteredResumes method to the Resume service and uses it in the Resume Book endpoint to support filtering through query parameters
* Adds a form next to the resume table that reloads the page with the appropriate query parameters for filtering on submit

I'm not that familiar with Go and SQL and not sure if this is the best way of implementing filtering, so let me know if there's a better way.

Other Questions:
1. The way I implemented filtering should be extensible enough to support filtering given an arbitrary model (such as User). So should it be extracted out and if so, to where? (I'm thinking like a utils folder, but don't see anything equivalent in the current structure)
2. Should the /resumemanager endpoint also support filtering?
3. In `controller/site/impl.go` (lines 478-481), I copied the available options for some of the fields (such as 2020-2025 for the graduation year) from the Resume Upload endpoint. I think these should be extracted out, but again not sure where they would go

ASankaran commented on service/resume/impl.go:
```diff
@@ -68,14 +79,41 @@ func (service *resumeImpl) validateResume(resume *model.Resume) error {
+				return nil, fmt.Errorf("invalid filter type for field '%v': %w", filterKey, err)
+			}
+			filters[filterKey] = filter
+			query += fmt.Sprintf(" AND %[1]v = :%[1]v", filterKey) // e.g. " AND username = :username"
```
At first glance this looked SQL-injectable, but it seem like this is safe since you're whitelisting `filterKey` against the fields in the resume model struct? Can you add a small comment about that to make sure that other reading this code in the future know that?

ASankaran commented:
> The way I implemented filtering should be extensible enough to support filtering given an arbitrary model (such as User). So should it be extracted out and if so, to where? (I'm thinking like a utils folder, but don't see anything equivalent in the current structure)

I agree that would be a good idea. Since the implementation looks SQL specific you could move the code to somewhere like `database/querybuilder/filter.go`.

> Should the /resumemanager endpoint also support filtering?

I hadn't thought of that originally, but I think that would be a useful feature.

> In `controller/site/impl.go` (lines 478-481), I copied the available options for some of the fields (such as 2020-2025 for the graduation year) from the Resume Upload endpoint. I think these should be extracted out, but again not sure where they would go

You could define them in`model/resume.go` as a constant or generate the months / years in a function. We do something kinda similar in `model/user.go` where we define valid values for some fields.

apappu9 commented:
- Extracted filtering functionality into separate file in `database/querybuilder/filter.go` and added clarifying comments
- Resume manager and API endpoint now also support filtering
- Valid resume field options are now specified in `model/resume.go`

Here's what the filter form currently looks like:

![image](https://user-images.githubusercontent.com/22779056/82288390-018eb680-9968-11ea-8a80-a332f34af811.png)

Let me know if anything should be modified

ASankaran commented:
Could you add a unit test which uses the new method which was added to the resume service? It can be similar to this unit test: https://github.com/ASankaran/core/blob/1e49dd765c94e72aa07ecb8b8d95c4b444b1413d/test/service/resume/resume_test.go#L28

apappu9 commented:
I added a couple basic unit tests, but let me know if I should make them more in depth.

Also, I'm unable to run the tests on my computer because they seem to be running assuming a production environment and trying to use google as the storage provider. Is this supposed to be case or am I missing something?

ASankaran commented on test/service/resume/resume_test.go:
```diff
@@ -147,3 +147,125 @@ func TestCreateAndApproveAndGetresume(t *testing.T) {
+		}
+	})
+
+	t.Run("Filter with no expected results", func(t *testing.T) {
```
nit: You didn't capitalize "Filter" in the other tests. Keep it lower case here as well for consistency.

ASankaran commented:
> I added a couple basic unit tests, but let me know if I should make them more in depth.
> 
> Also, I'm unable to run the tests on my computer because they seem to be running assuming a production environment and trying to use google as the storage provider. Is this supposed to be case or am I missing something?

Maybe I set something up wrong for unit-testing locally. I'll figure out why it's not working.

apappu9 commented:
Oh sorry, I was running the tests through my IDE instead of running `make test`. I'm now getting a different error about not finding the database `core-test`. Should I manually create this database or did I forget some step?

ASankaran commented:
Yeah, can you manually create the database, then run `make test-migration` and then `make test`. You essentially want to run the CI workflow defined here: https://github.com/ASankaran/core/blob/master/.github/workflows/test.yaml

Ill also update the development documentation to specify this in more details.

apappu9 commented:
Yep, that worked, thanks!
